lib/websocket.js

```diff
@@ -808,11 +808,10 @@ function receiverOnError(err) {
   const websocket = this[kWebSocket];
 
   websocket._socket.removeListener('data', socketOnData);
+  websocket._socket.resume();
 
-  websocket._readyState = WebSocket.CLOSING;
-  websocket._closeCode = err[kStatusCode];
+  websocket.close(err[kStatusCode]);
   websocket.emit('error', err);
-  websocket._socket.destroy();
```

Comment:
> And here, this no longer destroys the socket - instead it's now either closed cleanly when the remote end closes its end of the socket (by socketOnEnd) or it's destroyed after the close timeout. Right?

Correct.

> ...

> Not a disaster, but unnecessarily messy. I think this is easy to fix though: I think the sender should always call `socket.end()` after it sends a close frame, so that it half-closes the socket. It seems like that's always the right choice, even when there are no errors, becacuse we know for sure that we're not sending any data after the close frame. In that case, in the above example both peers would go to `socketOnEnd` and everything works fine.

> We still need to wait for the timeout if the remote peer never closes the socket regardless, but that's OK I think, that's when the timeout really is intended.

We can't do this because the data might be compressed. If a peer sends compressed data, a close frame, and then calls `socket.end()`, the `'end'` event on the socket of the receiving peer might be emitted before the data is decompressed because that is done asynchronously.

Reply:
It sounds like that's an issue with the receiving peer, rather than an issue with ending the socket, isn't it? Ending the socket should not necessarily immediately emit end on the websocket, if there's still asynchronous data being decompressed.

Does that problem exist today without this change? If a non-ws websocket client connects to a ws server, sends compressed data and then immediately shuts down the connection, what happens?

Reply:
> Does that problem exist today without this change? If a non-ws websocket client connects to a ws server, sends compressed data and then immediately shuts down the connection, what happens?

All data is read but the closing handshake will not work as expected. The close frame is not sent because the socket is closed before we had a change to write to it.

FWIW it was working as per your suggestion before support for permessage-deflate was added and then fixed with various patches.

Reply:
> The close frame is not sent because the socket is closed before we had a chance to write to it.

That's fixable though, I think... It's the receiving peer who is incorrectly closing the outgoing half of the socket that it needs to send back the close frame. It just needs to wait for its own processing before it does so. I think that's possible with the normal streams events.

Right now the logic (mostly in socketOnError) is just:

* Incoming socket end event -> immediate `receiver.end()` and outgoing `socket.end()`

What would happen if this was instead:

* Incoming socket end event -> `receiver.end()`
* Receiver finish event -> outgoing `socket.end()`

I.e. wait until the receiver is finished writing everything (including any outstanding close frames, if we've received a close frame) before the outgoing socket shuts down.

This only affects the compression case, where the first `.end()` call might not actually end the receiver immediately. I think the logic here to handle errors on the receiver will handle any cases where this goes wrong.

I think (I hope?) that after end() a writable will always either eventually emit 'finish' or 'error' so this ensures that one way or another we always clean up the socket, but also should mean that the data is fully decompressed and the close frame is sent before we close the outgoing half of the socket, if everything goes well (or the error will close the socket, if it doesn't).

Does that make sense? I'm not sure I've got the details 100% right, but I think this is close to something that would improve the compression behaviour in general, and there should be a route through here where we slightly decouple socket events from receiver events to allow this kind of async processing.

(This is getting a bit off topic here sorry, it seems that this is an existing bug even without this change, but it would be good to fix it up at the same time since they're related)

Reply:
> Right now the logic (mostly in socketOnError) is just:
> - Incoming socket end event -> immediate `receiver.end()` and outgoing `socket.end()`

Yes, because the other peer should not call `socket.end()` until it received a close frame. The `'end'` event should be unsolicited.

> What would happen if this was instead:

> - Incoming socket end event -> receiver.end()
> - Receiver finish event -> outgoing socket.end()

In this case the receiver might finish before we are done writing buffered (possibly to compress) data, so `socket.end()` might make the socket non-writable prematurely.

Reply:
> In this case the receiver might finish before we are done writing buffered (possibly to compress) data

Does the receiver call the write callback before decompression is finished? That shouldn't be called until all processing is completely finished. If that waits for decompression, then 'finish' will wait for decompression to complete, and everything is fine.

If it doesn't wait for decompression then this won't work, but in that case the receiver is breaking the stream invariants and all bets are off (and we should fix that too).

Reply:
Sorry, I know I'm making lots of suggestions and not actually helping much, let me know if you want me to look into some of these points separately, I'm happy to open PRs to look at some of these independently if that would be helpful.

Reply:
> Does the receiver call the write callback before decompression is finished? That shouldn't be called until all processing is completely finished. If that waits for decompression, then 'finish' will wait for decompression to complete, and everything is fine.

It does not, the callback is called when decompression is complete but the writable side of the WebSocket is completely independent from the readable side. Both the underlying socket and the WebSocket are duplex. If the other peer ends the writable side of the socket without receiving a close frame, the `Receiver` object would emit `'finish'` but this does not mean that the `Sender` object wrote a close frame and it we try to do that after `'finish'` -> `socket.end()` that would fail because the socket is no longer writable. Does it make sense?

Reply:
> the Receiver object would emit 'finish' but this does not mean that the Sender object wrote a close frame

Ah, yes, you're totally right.

As long as the remote peer sent a close frame, it seems that we will always call `sender.close()` before the receiver finishes (either because of 'conclude' or 'error'), which is good. Without compression everything is fine (as long as we get a close frame) but with compression `sender.close()` might buffer the close frame instead of sending it immediately.

That's manageable though! The change we're discussing here is to call `socket.end()` within `sender.close()`, so as long as close() is called, so we don't need to call socket.end() on 'finish' at all. Close() is always called before the receiver emits 'finish' (in the healthy case, where we get a close frame) and this will end the socket by itself.

The only remaining case is if the remote peer half-closes the connection, but doesn't send a close frame. I think right now in that case we emit 'close' 1005 and that's it? I.e. no error. That seems reasonable to me. We can do the same with this change by doing that on 'finish' if no close frame has been received.

This is confusing, so recapping all the changes here together:

* `socketOnEnd` should call `receiver.end()` instead of immediately killing the connection with `socket.end()`
* `sender.close()` should call `socket.end()` any time a close frame is sent (in the callback, once everything is written) to always start closing the connection.
* When the receiver emits 'finish', if no close frame was received then we need to call `websocket.close()` to make sure we clean everything up.

Does that work?

I think that would mean:

* We always clean up after socketOnEnd, because one of these happens:
  * Close frame before 'finish' -> sender.close(code) -> socket.end()
  * 'finish' with no close frame -> sender.close() -> socket.end()
  * 'error' in receiver -> sender.close() -> socket.end())
* We always wait for all buffered data to be sent & received in each of those cases, unless the socket itself errors
  * Wait for finish means all incoming data is processed
  * Wait for sender.close callback means all outgoing data was sent
  * `socketOnError` cleans up in every other case
* If the remote peer sends a close frame after we have sent a close frame, but then doesn't end the socket (e.g. because we both saw errors) then we will close the socket anyway (because sender.close() always closes the socket)

Reply:
Sure, that's reasonable. I'll look into it.

Without this, I think this PR still works OK. The only issue is the one edge case above where simultaneous errors can stop the handshake completing, so that you have to wait for the timeout, but it's not a big deal.

Reply:
I think I remember one of the edge cases. The HTTP server gave use a socket not allowed to be half-open, the server was provided by the user so we could not change the options. Anyway this should no longer be an issue in all supported Node.js versions.

Reply:
For completeness, with the current version of Node.js (v16.3.0), the server always gives us a socket allowed to be half-open, but the client, by default, gives us a socket not allowed to be half-open. We create the socket ourselves via the [`createConnection`](https://github.com/websockets/ws/blob/074e6a8be7275a69a407f6c1fa2270c754d2834b/lib/websocket.js#L519) option so we could create a socket with the `allowHalfOpen` option set to `true`, but we also allow users to use a custom [agent](https://nodejs.org/api/http.html#http_new_agent_options), so we would still need to handle the case where the client socket is not allowed to be half-open.

Reply:
See also https://github.com/nodejs/node/commit/c3b8e5014369561d86e32338f622034020c9c79f. It was backported only to Node.js v12.x

Reply:
That's really useful, thank you!

I think that was what I was missing, that makes it clearer why some of the current behaviour is useful. My comments above apply for the "half-open allowed" case, but I see how lots of that is more complicated when sockets might not allow that.

Even with the change in node that you linked, I agree we still have to deal with this, in both directions, because either a server or a client might provide a socket that we didn't initialize ourselves.

I think there might be an interesting improvement, where we just take over every socket when we receive it and change them to always allow half-open connections? I.e. `socket.allowHalfOpen = true`. That would make everything more consistent and easier to test - now all WS sockets now behave the same, and we can always properly complete the closing handshake with no risk of data loss. I don't think it would matter to code using ws - ws is going to take over `.end()` for the stream with or without this, it's just an implementation detail from their POV (AFAICT?).

That `allowHalfOpen` property is a duplex property (present at least back to node 0.10) that isn't used until the connection is closed, so this looks safe to me if it's set immediately. It's not formally documented though, I've filed an issue on node to confirm how this works and get that documented (if it is indeed safe to use it): https://github.com/nodejs/node/issues/38989/

I'll play with this and open a pull request in the next few days, once I've investigated everything more thoroughly, watch this space! Thanks for all the info, this is really interesting & useful.

Reply:
Before you spend time on investigating and preparing a PR I have to tell you that I have just remembered another reason why the "closing handshake" is implemented in its current form (that is `socket.end()` is called only after a close frame is both received and sent): 

Other clients/languages might not work like ws or support half-open sockets. For example see https://gist.github.com/lpinca/be56024eca9d4a1efc701364e6ec9513.

- Run it and open a browser (I tried with Chrome and Safari) then wait for some data to be buffered (I waited until ~20 MiB of data was buffered).
- Hit Ctrl-C. A close frame will be sent to the browser client.
- Wait until the data is written, then the node process will exit by its own.
- Check the file. The data includes a close frame and is framed but despite this, its size is smaller than the buffered data.
- The browser client simply drops all buffered data as soon as it detects that the socket is "half-closed".

Reply:
It might still worth investigating as removing `socket.end()` from the gist above does not work as expected on Chrome. However it kind of does on Safari.